### ISSUE TITLE
question about discriminator

### ISSUE THREAD

**Issue by lingdavid** [crux82/ganbert#5]
Hi, in discriminator code:  logit = tf.layers.dense(layer_hidden, (num_labels + 1)), why num_labels + 1?

> ############ Defining Discriminator ###########  
def discriminator(x, d_hidden_size, dkp, is_training, num_labels, num_hidden_discriminator=1, reuse=False):
    with tf.compat.v1.variable_scope('Discriminator', reuse=reuse):
        layer_hidden = tf.nn.dropout(x, keep_prob=dkp)
        for i in range(num_hidden_discriminator):
            layer_hidden = tf.layers.dense(layer_hidden, d_hidden_size)
            layer_hidden = tf.nn.leaky_relu(layer_hidden)
            layer_hidden = tf.nn.dropout(layer_hidden, keep_prob=dkp)
        flatten5 = layer_hidden
        logit = tf.layers.dense(layer_hidden, (num_labels + 1))
        prob = tf.nn.softmax(logit)
    return flatten5, logit, prob

**Comment by crux82** [comment]
The additional class is used to implement the Semi Supervised schema.
In this setting, the classifier is considered as a Discriminator (according to the formulation of a GAN) which is trained over a (k + 1)-class objective: “true” examples are classified in one of the target (1, ..., k) classes, while the "fake" samples derived from the generator are classified into the k + 1 class.

More details can be found here: https://www.aclweb.org/anthology/2020.acl-main.191.pdf

Hope it helps

Bests

Danilo

**Comment by lingdavid** [comment]
Thanks reply!
in create_model code:
`D_real_features, D_real_logits, D_real_prob = discriminator(output_layer, hidden_size, keep_prob, is_training,
                                                                num_labels, reuse=False)`
`logits = D_real_logits[:, 1:]`
Does the  D_real_logits[:, 0] mean the UNK label or the generated samples from generator?

**Comment by kamei86i** [comment]
So, with:

`logits = D_real_logits[:, 1:]`

we are selecting the logits relative to all the "real" categories, i.e., excluding UNK. So, yes, 

`logits = D_real_logits[:, 0]` 

should be the logit for UNK.

**Comment by Ashwin-Ramesh2607** [comment]
@kamei86i So if "0" index is for UNK, and [1: ] is for the real categories. Then which logit is for the classification of the "fake" sample.
